include/AudioBuffer.h

```diff
@@ -0,0 +1,308 @@
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_AUDIO_BUFFER_H
+#define LMMS_AUDIO_BUFFER_H
+
+#include <type_traits>
+#include <vector>
+
+#include "AudioData.h"
+#include "AudioPortsConfig.h"
+#include "LmmsTypes.h"
+#include "SampleFrame.h"
+
+namespace lmms
+{
+
+
+namespace detail {
+
+/**
+ * Metafunction to select the appropriate non-owning audio buffer view
+ * given the layout, sample type, and channel count
+ */
+template<AudioDataKind kind, bool interleaved, proc_ch_t channels, bool isConst>
+struct AudioDataViewTypeHelper
+{
+	static_assert(always_false_v<AudioDataViewTypeHelper<kind, interleaved, channels, isConst>>,
+		"Unsupported audio data type");
+};
+
+//! Non-interleaved specialization
+template<AudioDataKind kind, proc_ch_t channels, bool isConst>
+struct AudioDataViewTypeHelper<kind, false, channels, isConst>
+{
+	using type = SplitAudioData<
+		std::conditional_t<isConst, const GetAudioDataType<kind>, GetAudioDataType<kind>>,
+		channels>;
+};
+
+//! SampleFrame specialization
+template<proc_ch_t channels, bool isConst>
+struct AudioDataViewTypeHelper<AudioDataKind::SampleFrame, true, channels, isConst>
+{
+	static_assert(channels == 0 || channels == 2,
+		"Plugins using SampleFrame buffers must have exactly 0 or 2 inputs or outputs");
+	using type = std::conditional_t<isConst, std::span<const SampleFrame>, std::span<SampleFrame>>;
+};
+
+} // namespace detail
+
+
+//! Metafunction to select the appropriate non-owning audio buffer view
+template<AudioPortsConfig config, bool isOutput, bool isConst>
+using AudioDataViewType = typename detail::AudioDataViewTypeHelper<
+	config.kind, config.interleaved, (isOutput ? config.outputs : config.inputs), isConst>::type;
+
+
+namespace detail {
+
+//! Interface for accessing input/output audio buffers
+template<AudioPortsConfig config, bool inplace = config.inplace>
+class AudioBuffer;
+
+//! Dynamically in-place specialization
+template<AudioPortsConfig config>
+class AudioBuffer<config, false>
+{
+public:
+	virtual ~AudioBuffer() = default;
+
+	virtual auto inputBuffer() -> AudioDataViewType<config, false, false> = 0;
+	virtual auto outputBuffer() -> AudioDataViewType<config, true, false> = 0;
+	virtual auto frames() const -> fpp_t = 0;
+	virtual void updateBuffers(proc_ch_t channelsIn, proc_ch_t channelsOut, f_cnt_t frames) = 0;
```

Comment:
Are these `virtual` functions called in the audio processing itself?

Reply:
`inputBuffer` and `outputBuffer` are, yes. Though for `DefaultAudioBuffer`, they are marked `final` to let the compiler optimize away the vtable lookup.